motionplan/planManager.go

```diff
@@ -0,0 +1,428 @@
+	alg, ok := planningOpts["planning_alg"]
+	if ok {
+		planAlg, ok = alg.(string)
+		if !ok {
+			return nil, errors.New("could not interpret planning_alg field as string")
+		}
+		switch planAlg {
+		// TODO(pl): make these consts
+		case "cbirrt":
+			opt.PlannerConstructor = newCBiRRTMotionPlanner
+		case "rrtstar":
+			// no motion profiles for RRT*
+			opt.PlannerConstructor = newRRTStarConnectMotionPlanner
+			// TODO(pl): more logic for RRT*?
+			return opt, nil
+		default:
+			// use default, already set
+		}
+	}
+
+	switch motionProfile {
+	case LinearMotionProfile:
+		// Linear constraints
+		linTol, ok := planningOpts["line_tolerance"].(float64)
+		if !ok {
+			// Default
+			linTol = defaultLinearDeviation
+		}
+		orientTol, ok := planningOpts["orient_tolerance"].(float64)
+		if !ok {
+			// Default
+			orientTol = defaultLinearDeviation
+		}
+		constraint, pathDist := NewAbsoluteLinearInterpolatingConstraint(from, to, linTol, orientTol)
+		opt.AddConstraint(defaultLinearConstraintName, constraint)
+		opt.pathDist = pathDist
+	case PseudolinearMotionProfile:
+		tolerance, ok := planningOpts["tolerance"].(float64)
+		if !ok {
+			// Default
+			tolerance = defaultPseudolinearTolerance
+		}
+		constraint, pathDist := NewProportionalLinearInterpolatingConstraint(from, to, tolerance)
+		opt.AddConstraint(defaultPseudolinearConstraintName, constraint)
+		opt.pathDist = pathDist
+	case OrientationMotionProfile:
+		tolerance, ok := planningOpts["tolerance"].(float64)
+		if !ok {
+			// Default
+			tolerance = defaultOrientationDeviation
+		}
+		constraint, pathDist := NewSlerpOrientationConstraint(from, to, tolerance)
+		opt.AddConstraint(defaultOrientationConstraintName, constraint)
+		opt.pathDist = pathDist
+	case PositionOnlyMotionProfile:
+		opt.SetMetric(NewPositionOnlyMetric())
+	case FreeMotionProfile:
+		// No restrictions on motion
+		fallthrough
+	default:
+		if planAlg == "" {
+			// set up deep copy for fallback
+			try1 := deepAtomicCopyMap(planningOpts)
+			// No need to generate tons more IK solutions when the first alg will do it
+			opt.MaxSolutions = defaultFallbackIK
+
+			// extra timeout to account for IK
+			try1["timeout"] = defaultFallbackTimeout
+			try1["planning_alg"] = "rrtstar"
+			try1Opt, err := mp.plannerSetupFromMoveRequest(from, to, seedMap, worldState, try1)
+			if err != nil {
+				return nil, err
+			}
+
+			try1Opt.Fallback = opt
+			opt = try1Opt
+		}
+	}
+	return opt, nil
+}
+
+// check whether the solution is within some amount of the optimal.
+func goodPlan(pr *rrtPlanReturn, opt *plannerOptions) (bool, float64) {
+	solutionCost := 0.
+	if pr.steps != nil {
+		if pr.optimal <= 0 {
+			return true, solutionCost
+		}
+		solutionCost = EvaluatePlan(pr, opt)
+		if solutionCost < pr.optimal*defaultOptimalityMultiple {
+			return true, solutionCost
+		}
+	}
+
+	return false, solutionCost
+}
+
+// Copy any atomic values.
+func deepAtomicCopyMap(opt map[string]interface{}) map[string]interface{} {
```

Comment:
[nit] this could potentially be in a utils folder, but I don't think its a big deal either way

Reply:
This should be used with care, as it'll only deep copy atomic values; if it's public and in utils, someone will eventually use it to copy a map containing a pointer to something, and then things will break.